ramp4-pcar4/ramp4-pcar4
# Add CRUD methods to legend

## Closes #1222, Closes #1051
### Refers #1046, #1133, #938

## Changes
- Added the legend api CRUD suite as per discussion #1217 (with some tweaks)
- Added a new minimal `legend` sample to demos that runs a few test cases and dumps the output to the console
    - Note that for the sake of this PR, I added a `index-legend` to the public folder so it can be accessed in the PR build
    - I will remove this public sample before merging
- Fixed bug where items from the legend fixture config were being added in reverse order
- Fixed bug where MIL sublayers will ignore layer state and always set visibility to `true` and opacity to `1`
- Fixed bug where MIL sublayers were using their own `TreeNode` instead of the one created by the MIL root
- Added support for layer tree-grow where adding layers will have their layer tree mapped into legend items
- Removed `<KeepAlive>` tags around the legend item component because it was giving issues with the checkbox reactivity
- Added wizard appbar button to fully support wizard decoupling

## Notes
- I've noticed many cases where fixture components require access to its fixture api and so when the component is mounted/created, it has to manually fetch the api with `this.$iApi.fixture.get('fixture-name')`
    -  Just like how the panel instance is provided to all these components as a prop, maybe we can also provide the fixture api (if it has one)?
    - If this is a good idea, I can create a new discussion post + issue to look into this
- The new `addLayerItem` method now maps the layer tree into a legend tree
    - I think this donethanks' #1051 and handles parts of discussion #1133,
    - Would appreciate @yileifeng's and @james-rae's feedback on whether this can be closed
    - I also tried to reproduce #1046 but it seems like the MIL `onLoadActions` throws an error with this layer, so this issue can remain open

## Testing

[Public Legend Sample](https://ramp4-pcar4.github.io/ramp4-pcar4/1222/index-legend.html) - will be removed before this PR is merged
- The console should shows the outputs of various test calls to the api
- Use `window.lApi` to access the legend api and call its methods
- The legend should function the same as before
- Try adding `Group1` from this [MapImageLayer](https://section917.canadacentral.cloudapp.azure.com/arcgis/rest/services/TestData/Nest/MapServer) - the layer tree structure in the server should be mirrored in the legend


---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/1281)
